src/ParallelAlgorithms/LinearSolver/ConjugateGradient/ElementActions.hpp

```diff
@@ -232,19 +249,37 @@ struct UpdateOperand {
       db::add_tag_prefix<LinearSolver::Tags::Residual, fields_tag>;
 
  public:
-  template <
-      typename ParallelComponent, typename DbTagsList, typename Metavariables,
-      typename ArrayIndex, typename DataBox = db::DataBox<DbTagsList>,
-      Requires<db::tag_is_retrievable_v<fields_tag, DataBox> and
-               db::tag_is_retrievable_v<
-                   LinearSolver::Tags::HasConverged<OptionsGroup>, DataBox> and
-               db::tag_is_retrievable_v<
-                   LinearSolver::Tags::IterationId<OptionsGroup>, DataBox> and
-               db::tag_is_retrievable_v<residual_tag, DataBox>> = nullptr>
-  static auto apply(db::DataBox<DbTagsList>& box,
-                    Parallel::GlobalCache<Metavariables>& cache,
-                    const ArrayIndex& array_index, const double res_ratio,
-                    const Convergence::HasConverged& has_converged) noexcept {
+  using inbox_tags =
+      tmpl::list<Tags::ResidualRatioAndHasConverged<OptionsGroup>>;
```

Comment:
The Krylov-subspace solvers (GMRES and CG) work by computing global inner products to build up a basis of the Krylov subspace (that's just K_n(A,b)={b, A b, A^2 b, ..., A^(n-1) b}). The global inner products are needed to orthogonalize the basis. It's a nice side-effect that the global reductions also yield the residual magnitude, so it can be used to terminate the algorithm when it falls below a tolerance.

On the other hand, the linear solvers I call "asynchronous" don't need global synchronization points to work. For example, the Richardson solver just iterates every element independently, and the Schwarz solver only communicates data on overlap regions with neighboring elements. Therefore, they only dispatch reductions to _observe_ the residual magnitude, but don't wait for broadcasts.

Note that this issue with global synchronization points is one of the reasons why we use a GMRES solver with a (Multigrid-)Schwarz preconditioner for elliptic solves: The outer GMRES solver only runs for ~10 iterations, so the global synchronizations are not a problem, whereas the Schwarz preconditioner that runs in every GMRES step does all the heavy lifting in parallel.

Reply:
Aha, got it. Thanks for the explanation.